src/iterators/idlist_iterator.c

```diff
@@ -70,62 +69,112 @@ static IteratorStatus IL_SkipTo(QueryIterator *base, t_docId docId) {
     did = it->docIds[++i];
   }
   it->offset = i + 1;
-  if (it->offset >= it->size) {
-    setEof(it, true);
-  }
-
-  it->base.current->docId = it->base.lastDocId = did;
-  return docId == did ? ITERATOR_OK : ITERATOR_NOTFOUND;
+  base->current->docId = base->lastDocId = did;
+  return docId == base->lastDocId ? ITERATOR_OK : ITERATOR_NOTFOUND;
 }
 
 /* release the iterator's context and free everything needed */
 static void IL_Free(QueryIterator *self) {
   IdListIterator *it = (IdListIterator *)self;
-  IndexResult_Free(it->base.current);
-  if (it->docIds) {
-      rm_free(it->docIds);
-  }
+  IndexResult_Free(self->current);
+  rm_free(it->docIds);
   rm_free(self);
 }
- 
+
 static int cmp_docids(const void *p1, const void *p2) {
   const t_docId *d1 = p1, *d2 = p2;
-
-  return (int)(*d1 - *d2);
+  if (*d1 < *d2) return -1;
+  if (*d1 > *d2) return 1;
+  return 0;
 }
 
 static void IL_Rewind(QueryIterator *base) {
   IdListIterator *il = (IdListIterator *)base;
-  setEof(il, false);
-  il->base.lastDocId = 0;
-  il->base.current->docId = 0;
+  setEof(base, false);
+  base->lastDocId = 0;
+  base->current->docId = 0;
   il->offset = 0;
 }
- 
+
 QueryIterator *IT_V2(NewIdListIterator) (t_docId *ids, t_offset num, double weight) {
   // Assume the ids are not null and num > 0 otherwise these Iterator would not be created, avoid validation
   // first sort the ids, so the caller will not have to deal with it
-  
-  qsort(ids, (size_t)num, sizeof(t_docId), cmp_docids);
-
   IdListIterator *it = rm_new(IdListIterator);
-
   it->size = num;
-  it->docIds = rm_calloc(num, sizeof(t_docId));
-  if (num > 0) memcpy(it->docIds, ids, num * sizeof(t_docId));
-  setEof(it, false);
-  it->base.current = NewVirtualResult(weight, RS_FIELDMASK_ALL);
-  it->base.lastDocId = 0;
-
+  it->docIds = ids;
   it->offset = 0;
-
   QueryIterator *ret = &it->base;
+
+  setEof(ret, false);
+  ret->current = NewVirtualResult(weight, RS_FIELDMASK_ALL);
+  ret->lastDocId = 0;
   ret->type = ID_LIST_ITERATOR;
   ret->NumEstimated = IL_NumEstimated;
   ret->Free = IL_Free;
   ret->Read = IL_Read;
   ret->SkipTo = IL_SkipTo;
   ret->Rewind = IL_Rewind;
+  return ret;
+}
 
+static void SetYield(QueryIterator *base, double value) {
+  base->current->num.value = value;
+  if (!base->current->metrics) {
+    ResultMetrics_Add(base->current, NULL, RS_NumVal(value));
+  } else {
+    ResultMetrics_UpdateDouble(base->current, NULL, value);
```

Comment:
Sadly, we can't do this optimization this way.
The metric and `RSValue` may be shared at this point (after we previously returned them) is the caller took a reference to the result. This is why we must allocate a new RSValue every time.
We can have an optimization that checks if we have the only reference to this value, and then reuse it